--- hpctoolkit.ORG/src/tool/hpcrun-flat/monitor_preload.c	2015-10-01 10:44:10.062344065 +0530
+++ hpctoolkit/src/tool/hpcrun-flat/monitor_preload.c	2015-10-05 10:32:49.473741014 +0530
@@ -108,12 +108,25 @@
   fini_library();
 }
 
-
+#define PROC_NAME_LEN  2048
+char buf[PROC_NAME_LEN];
 void 
-monitor_init_process(const char* name, int* argc, char** argv, 
-		     const unsigned pid)
+monitor_init_process(int* argc, char** argv, void* data)
 {
-  hpcrun_cmd = name;  /* command is also in /proc/pid/cmdline */
+  // FIXME: if the process fork()s before main, then argc and argv
+  // will be NULL in the child here.  MPT on CNL does this.
+  hpcrun_cmd = "unknown";
+  if (argv != NULL && argv[0] != NULL) {
+    hpcrun_cmd = argv[0];
+  }
+  else {
+    int len = readlink("/proc/self/exe", buf, PROC_NAME_LEN - 1);
+    if (len > 1) {
+      buf[len] = 0;
+      hpcrun_cmd = buf;
+    }
+  }
+
   if (opt_debug >= 1) {
     fprintf(stderr, "Init process callback from monitor received\n");
   }
